presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -87,150 +92,189 @@ private Rewriter(PlanNodeIdAllocator idAllocator, Metadata metadata, Session ses
+        private static boolean isEqualRange(TupleDomain<Symbol> tupleDomain, Symbol symbol, long upperBound)
         {
-            Set<Symbol> windowFunctionSymbols = node.getWindowFunctions().keySet();
-            Sets.SetView<Symbol> commonSymbols = Sets.intersection(DependencyExtractor.extractUnique(filterPredicate), windowFunctionSymbols);
-            return !commonSymbols.isEmpty();
+            return tupleDomain.getDomains().get(symbol).getRanges().equals(SortedRangeSet.of(Range.lessThanOrEqual(upperBound)));
         }
 
-        @Override
-        public PlanNode visitLimit(LimitNode node, RewriteContext<Constraint> context)
+        private static Optional<Integer> extractUpperBound(TupleDomain<Symbol> tupleDomain, Symbol symbol)
         {
-            // Operators can handle MAX_VALUE rows per page, so do not optimize if count is greater than this value
-            if (node.getCount() >= Integer.MAX_VALUE) {
-                return context.defaultRewrite(node);
+            if (tupleDomain.isNone()) {
+                return Optional.empty();
+            }
+
+            Domain rowNumberDomain = tupleDomain.getDomains().get(symbol);
+            if (rowNumberDomain == null) {
+                return Optional.empty();
             }
-            Constraint constraint = new Constraint(Optional.of((int) node.getCount()), Optional.empty());
-            PlanNode rewrittenSource = context.rewrite(node.getSource(), constraint);
 
-            if (rewrittenSource != node.getSource()) {
-                return new LimitNode(idAllocator.getNextId(), rewrittenSource, node.getCount());
+            SortedRangeSet ranges = rowNumberDomain.getRanges();
+            if (ranges.isAll() || ranges.isNone() || ranges.getRangeCount() <= 0) {
+                return Optional.empty();
             }
 
-            return context.defaultRewrite(node);
+            Range span = ranges.getSpan();
+
+            if (span.getHigh().isUpperUnbounded()) {
+                return Optional.empty();
+            }
+
+            verify(rowNumberDomain.getType().equals(Long.class));
+            long upperBound = (Long) span.getHigh().getValue();
+            if (span.getHigh().getBound() == BELOW) {
+                upperBound--;
+            }
+
+            if (upperBound > Integer.MAX_VALUE) {
+                return Optional.empty();
+            }
+            return Optional.of(Ints.checkedCast(upperBound));
         }
 
-        @Override
-        public PlanNode visitFilter(FilterNode node, RewriteContext<Constraint> context)
+        private static RowNumberNode mergeLimit(RowNumberNode node, int newRowCountPerPartition)
         {
-            PlanNode rewrittenSource = context.rewrite(node.getSource(), new Constraint(Optional.empty(), Optional.of(node.getPredicate())));
-            return context.replaceChildren(node, ImmutableList.of(rewrittenSource));
+            int rowCountPerPartition = newRowCountPerPartition;
```

Comment:
don't need rowCountPerPartition variable, you can directly just overwrite newRowCountPerPartition